### PR TITLE
use property based tests for BFieldElement

### PR DESCRIPTION
**Pull request by jan-ferdinand**
Refactor and clean up some BFieldElement tests so that they only do one thing. Use an actual property testing framework for property based tests.

### CODE REVIEW

File: twenty-first/src/shared_math/b_field_element.rs [hunk 1]
```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
...
 
-    #[test]
-    fn simple_value_test() {
-        let zero: BFieldElement = BFieldElement::new(0);
-        assert_eq!(0, zero.value());
-        let one: BFieldElement = BFieldElement::one();
-        assert_eq!(1, one.value());
+    proptest! {
+        #[test]
+        fn multiplication_is_associative(
```
Comment (Sword-Smith) [inline twenty-first/src/shared_math/b_field_element.rs:625]: Nice mathy tests!

File: twenty-first/src/shared_math/b_field_element.rs [hunk 2]
```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
...
+            prop_assert!((bfe / bfe).is_one());
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn values_larger_than_modulus_are_handled_correctly(large_value in BFieldElement::P..) {
+            let bfe = BFieldElement::new(large_value);
+            let expected_value = large_value - BFieldElement::P;
+            prop_assert_eq!(expected_value, bfe.value());
```
Comment (Sword-Smith) [inline twenty-first/src/shared_math/b_field_element.rs:670]: Wouldn't this test be marginally more readable as a one-liner? 

```rust
prop_assert_eq!(large_value - BFieldElement::P, BFieldElement::new(large_value).value());
```

File: twenty-first/src/shared_math/b_field_element.rs [hunk 3]
```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
...
         }
+    }
 
-        let empty_inversion = BFieldElement::batch_inversion(vec![]);
-        assert!(empty_inversion.is_empty());
+    proptest! {
+        #[test]
+        fn value_is_preserved(value in 0..BFieldElement::P) {
+            let bfe = BFieldElement::new(value);
+            prop_assert_eq!(value, bfe.value());
```
Comment (Sword-Smith) [inline twenty-first/src/shared_math/b_field_element.rs:741]: This is also marginally more readable as a one-liner, I think.

File: twenty-first/src/shared_math/b_field_element.rs [hunk 4]
```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
...
     }
 
-    #[test]
-    fn inversion_property_based_test() {
-        let elements: Vec<BFieldElement> = random_elements(30);
+    proptest! {
+        #[test]
+        fn lift_then_unlift_preserves_element(bfe: BFieldElement) {
+            let maybe_same_bfe = bfe.lift().unlift();
+            prop_assert_eq!(Some(bfe), maybe_same_bfe);
```
Comment (Sword-Smith) [inline twenty-first/src/shared_math/b_field_element.rs:760]: one-liner?

File: twenty-first/src/shared_math/b_field_element.rs [hunk 5]
```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
...
+        assert_eq!("-15", format!("{minus_fifteen}"));
     }
 
     #[test]
-    fn lift_property_test() {
-        let elements: Vec<BFieldElement> = random_elements(100);
-        for element in elements {
-            assert_eq!(Some(element), element.lift().unlift());
-        }
+    fn zero_is_zero() {
```
Comment (Sword-Smith) [inline twenty-first/src/shared_math/b_field_element.rs:687]: `zero_is_zero` and `one_is_one` should also check that `one.is_zero()` and `zero.is_one()` returns false, and that an arbitrary value (can be hardcoded) returns false on `is_zero()` and `is_one()`.

File: twenty-first/src/shared_math/b_field_element.rs [hunk 6]
```diff
@@ -552,286 +552,251 @@ mod b_prime_field_element_test {
...
+        let zero = BFieldElement::zero();
+        assert!(zero.is_zero());
     }
 
     #[test]
-    fn next_and_previous_test() {
```
Comment (Sword-Smith) [inline twenty-first/src/shared_math/b_field_element.rs:674]: This test checked that `increment` and `decrement` handled the corner-cases correctly wrapping around the prime. I think your property-based tests don't hit those corner-cases.

**Review by Sword-Smith — CHANGES_REQUESTED** [review]

**Review by Sword-Smith — CHANGES_REQUESTED** [review]
Running `cargo t` produced a file that is not committed to git. What should we do with this file, and does it somehow indicate an error? All tests have always passed as I ran them on my machine.

```
$ cat twenty-first/proptest-regressions/shared_math/b_field_element.txt 
# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc 475795502dfa9fb101aaf413911719f066f2e8698d634e9b11fc594ef3d8f71d # shrinks to byte_array = [1, 0, 0, 0, 255, 255, 255, 255]
```

File: twenty-first/src/shared_math/b_field_element.rs (hunk 5 above)
Comment (jan-ferdinand) [reply twenty-first/src/shared_math/b_field_element.rs:687]: I've added the tests `not_zero_is_nonzero` and `not_one_is_not_one` that cover these properties more fundamentally, and as their own tests, not muddled in with an existing test. Also note the existence of test `one_unequal_zero`.

File: twenty-first/src/shared_math/b_field_element.rs (hunk 2 above)
Comment (jan-ferdinand) [reply twenty-first/src/shared_math/b_field_element.rs:670]: I disagree with this one: the suggested one-liner is (a) more difficult to digest and (b) exceeds the line length limit.

File: twenty-first/src/shared_math/b_field_element.rs (hunk 6 above)
Comment (jan-ferdinand) [reply twenty-first/src/shared_math/b_field_element.rs:674]: You are right, they corners are hit too rarely. I've added the tests `increment_overflow` and `decrement_underflow` for the two corner cases.

**Comment by jan-ferdinand** [comment]
> Does [the file] somehow indicate an error?

This seems to be a [known bug](https://github.com/proptest-rs/proptest/issues/263) of proptest for tests marked as `should_panic`. Therefore, it does not indicate an error in this specific case, but it is also quite misleading.

> What should we do with this file?

Good question. I'm not sure. I think committing it once an actual violated property has been identified makes good sense. If my understanding of how proptest works is correct, committing it now also does not hurt, but my understanding of proptest's internals is only starting to develop.

File: twenty-first/src/shared_math/b_field_element.rs (hunk 6 above)
Comment (Sword-Smith) [reply twenty-first/src/shared_math/b_field_element.rs:674]: Great. I don't like the names, though. Maybe `increment_wrap`, `decrement_wrap` instead? Underflow refers to a floating point becoming zero, not to a wrap around from zero to the highest possible value.

**Review by Sword-Smith — APPROVED** [review]

**Review by Sword-Smith — CHANGES_REQUESTED** [review]
The generation of the proptest error file seems like a regression to me. We could consider ignoring it in `.gitignore`. That way is doesn't pollute your git interface/command line output with problematic uncommitted files.

**Comment by Sword-Smith** [comment]
Another solution would be to let failing tests be regular tests, and not proptests. Concretely that would mean that `byte_array_outside_range_cannot_be_converted` shouldn't be a proptest.

File: twenty-first/src/shared_math/b_field_element.rs (hunk 6 above)
Comment (jan-ferdinand) [reply twenty-first/src/shared_math/b_field_element.rs:674]: I've made it even more explicit with `incrementing_max_value_wraps_around` and `decrementing_min_value_wraps_around`.

**Comment by jan-ferdinand** [comment]
The correct solution is to not implement `From<[u8; 8]>` but `TryFrom<[u8; 8]>`. The documentation of the `From` trait literally says:

> **This trait must not fail.** The `From` trait is intended for perfect conversions. If the conversion can fail or is not perfect, use `TryFrom`.

In this case, the test doesn't have to panic, but the returned `Result<_, _>` can be checked for `is_err()`.

**Comment by Sword-Smith** [comment]
> The correct solution is to not implement `From<[u8; 8]>` but `TryFrom<[u8; 8]>`. The documentation of the `From` trait literally says:
> 
> > **This trait must not fail.** The `From` trait is intended for perfect conversions. If the conversion can fail or is not perfect, use `TryFrom`.
> 
> In this case, the test doesn't have to panic, but the returned `Result<_, _>` can be checked for `is_err()`.

Good point. I have no problem with replacing the `From<[u8; 8]>` implementation with a `TryFrom` implementation. Perhaps we should get the thumbs up from @aszepieniec before doing that though, since, AFAIK, he introduced this trait implementation when implementing Tip5.

**Comment by jan-ferdinand** [comment]
With #146 merged, no spurious proptest failure file is being generated, resolving (as far as I can see) the last remaining change request.
